File: OnionMan_Unity/Assets/_Assets/Network/EncodingUtility.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnityEngine;

namespace OnionMan.Network
{
    public class EncodingUtility : MonoBehaviour
    {
        public static IEnumerable<byte> Encode<T>(T value)
        {
            switch (value)
            {
                #region Base Types
                //Misc
                case string stringValue:
                    return Encoding.UTF8.GetBytes(stringValue);
                case bool boolValue:
                    return BitConverter.GetBytes(boolValue);
                case char charValue:
                    return BitConverter.GetBytes(charValue);

                //Numbers
                case double doubleValue:
                    return BitConverter.GetBytes(doubleValue);
                case float floatValue:
                    return BitConverter.GetBytes(floatValue);
                case int intValue:
                    return BitConverter.GetBytes(intValue);
                case long longValue:
                    return BitConverter.GetBytes(longValue);
                case short shortValue:
                    return BitConverter.GetBytes(shortValue);
                case uint uintValue:
                    return BitConverter.GetBytes(uintValue);
                case ulong ulongValue:
                    return BitConverter.GetBytes(ulongValue);
                case ushort ushortValue:
                    return BitConverter.GetBytes(ushortValue);
                #endregion
                #region Derived Types
                //Vectors
                case Vector3 vector3Value:
                    return Encode(vector3Value.x)
                        .Concat(Encode(vector3Value.y))
                        .Concat(Encode(vector3Value.z));

                case Vector3Int vector3IntValue:
                    return Encode(vector3IntValue.x)
                        .Concat(Encode(vector3IntValue.y))
                        .Concat(Encode(vector3IntValue.z));

                case Vector2 vector2Value:
                    return Encode(vector2Value.x)
                        .Concat(Encode(vector2Value.y));

                case Vector2Int vector2IntValue:
                    return Encode(vector2IntValue.x)
                        .Concat(Encode(vector2IntValue.y));

                case Quaternion quaternionValue:
                    return Encode(quaternionValue.x)
                        .Concat(Encode(quaternionValue.y))
                        .Concat(Encode(quaternionValue.z))
                        .Concat(Encode(quaternionValue.w));
                #endregion
                default:
                    throw new NotImplementedException($"The type {typeof(T)} cannot be encoded yet, u looser");
            }
        }

        public static void PutEncodedValueInBuffer<T>(T value, byte[] buffer, ref int offset)
        {
            switch (value)
            {
                #region Base Types
                //Misc
                case string stringValue:
                    PutToBuffer(buffer, Encoding.UTF8.GetBytes(stringValue), ref offset);
                    break;
                case bool boolValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(boolValue), ref offset);
                    break;
                case char charValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(charValue), ref offset);
                    break;

                //Numbers
                case double doubleValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(doubleValue), ref offset);
                    break;
                case float floatValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(floatValue), ref offset);
                    break;
                case int intValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(intValue), ref offset);
                    break;
                case long longValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(longValue), ref offset);
                    break;
                case short shortValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(shortValue), ref offset);
                    break;
                case uint uintValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(uintValue), ref offset);
                    break;
                case ulong ulongValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(ulongValue), ref offset);
                    break;
                case ushort ushortValue:
                    PutToBuffer(buffer, BitConverter.GetBytes(ushortValue), ref offset);
                    break;
                #endregion
                #region Derived Types
                //Vectors
                case Vector3 vector3Value:
                    PutEncodedValueInBuffer(vector3Value.x, buffer, ref offset);
                    PutEncodedValueInBuffer(vector3Value.y, buffer, ref offset);
                    PutEncodedValueInBuffer(vector3Value.z, buffer, ref offset);
                    break;

                case Vector3Int vector3IntValue:
                    PutEncodedValueInBuffer(vector3IntValue.x, buffer, ref offset);
                    PutEncodedValueInBuffer(vector3IntValue.y, buffer, ref offset);
                    PutEncodedValueInBuffer(vector3IntValue.z, buffer, ref offset);
                    break;

                case Vector2 vector2Value:
                    PutEncodedValueInBuffer(vector2Value.x, buffer, ref offset);
                    PutEncodedValueInBuffer(vector2Value.y, buffer, ref offset);
                    break;

                case Vector2Int vector2IntValue:
                    PutEncodedValueInBuffer(vector2IntValue.x, buffer, ref offset);
                    PutEncodedValueInBuffer(vector2IntValue.y, buffer, ref offset);
                    break;

                case Quaternion quaternionValue:
                    PutEncodedValueInBuffer(quaternionValue.x, buffer, ref offset);
                    PutEncodedValueInBuffer(quaternionValue.y, buffer, ref offset);
                    PutEncodedValueInBuffer(quaternionValue.z, buffer, ref offset);
                    PutEncodedValueInBuffer(quaternionValue.w, buffer, ref offset);
                    break;
                #endregion
                default:
                    throw new NotImplementedException($"The type {typeof(T)} cannot be encoded yet, u looser");
            }
        }

        public static T Decode<T>(byte[] bytes, ref int offset, int size = -1)
        {
            int initialOffset = offset;
            // string are too specific
            if (typeof(T) == typeof(string))
            {
                if (size == -1)
                {
                    throw new ArgumentException($"In order to decode a string, the argument 'size' must be set");
                }
                object st = Encoding.UTF8.GetString(bytes, offset, size);
                offset += size;
                return (T)st;
            }

            object decodedValue;
            T defaultObj = GetDefaultObjOfType<T>();

            switch (defaultObj)
            {
                #region Base Types
                //Misc
                case bool:
                    decodedValue = BitConverter.ToBoolean(bytes, offset);
                    offset += sizeof(bool);
                    break;
                case char:
                    decodedValue = BitConverter.ToChar(bytes, offset);
                    offset += sizeof(char);
                    break;

                //Numbers
                case double:
                    decodedValue = BitConverter.ToDouble(bytes, offset);
                    offset += sizeof(double);
                    break;
                case float:
                    decodedValue = BitConverter.ToSingle(bytes, offset);
                    offset += sizeof(float);
                    break;
                case int:
                    decodedValue = BitConverter.ToInt32(bytes, offset);
                    offset += sizeof(int);
                    break;
                case long:
                    decodedValue = BitConverter.ToInt64(bytes, offset);
                    offset += sizeof(long);
                    break;
                case short:
                    decodedValue = BitConverter.ToInt16(bytes, offset);
                    offset += sizeof(short);
                    break;
                case uint:
                    decodedValue = BitConverter.ToUInt32(bytes, offset);
                    offset += sizeof(uint);
                    break;
                case ulong:
                    decodedValue = BitConverter.ToUInt64(bytes, offset);
                    offset += sizeof(ulong);
                    break;
                case ushort:
                    decodedValue = BitConverter.ToUInt16(bytes, offset); 
                    offset += sizeof(ushort);
                    break;
                #endregion
                #region Derived Types
                //Vectors
                case Vector3:
                    decodedValue = new Vector3(
                        Decode<float>(bytes, ref offset),
                        Decode<float>(bytes, ref offset),
                        Decode<float>(bytes, ref offset));
                    break;

                case Vector3Int:
                    decodedValue = new Vector3Int(
                        Decode<int>(bytes, ref offset),
                        Decode<int>(bytes, ref offset),
                        Decode<int>(bytes, ref offset));
                    break;

                case Vector2:
                    decodedValue = new Vector2(
                        Decode<float>(bytes, ref offset),
                        Decode<float>(bytes, ref offset));
                    break;

                case Vector2Int:
                    decodedValue = new Vector2Int(
                        Decode<int>(bytes, ref offset),
                        Decode<int>(bytes, ref offset));
                    break;

                case Quaternion:
                    decodedValue = new Quaternion(
                        Decode<float>(bytes, ref offset),
                        Decode<float>(bytes, ref offset),
                        Decode<float>(bytes, ref offset),
                        Decode<float>(bytes, ref offset));
                    break;
                #endregion
                default:
                    throw new NotImplementedException($"The type {typeof(T)} cannot be decoded yet");
            }

            if (size != -1 && offset - initialOffset != size)
            {
                Debug.LogError("Offset Overflow !");
            }
            return (T)decodedValue;
        }

        public static int GetEncodedSize<T>(T value = default)
        {
            // string are too specific
            if (typeof(T) == typeof(string))
            {
                if (value == null)
                {
                    throw new ArgumentException($"In order to get a string size, it shouldn't be null or default");
                }
                return Encode(value).Count();
            }

            T defaultObj = GetDefaultObjOfType<T>();

            switch (defaultObj)
            {
                #region Base Types
                //Misc
                case bool:
                    return sizeof(bool);
                case char:
                    return sizeof(char);

                //Numbers
                case double:
                    return sizeof(double);
                case float:
                    return sizeof(float);
                case int:
                    return sizeof(int);
                case long:
                    return sizeof(long);
                case short:
                    return sizeof(short);
                case uint:
                    return sizeof(uint);
                case ulong:
                    return sizeof(ulong);
                case ushort:
                    return sizeof(ushort);
                #endregion
                #region Derived Types
                //Vectors
                case Vector3:
                    return sizeof(float) * 3;

                case Vector3Int:
                    return sizeof(int) * 3;

                case Vector2:
                    return sizeof(float) * 2;

                case Vector2Int:
                    return sizeof(int) * 2;

                case Quaternion:
                    return sizeof(float) * 4;
                #endregion
                default:
                    throw new NotImplementedException($"The size of {typeof(T)} cannot be found yet");
            }
        }
        public static bool HasFixedEncodedSize<T>()
        {
            // string are too specific
            if (typeof(T) == typeof(string))
            {
                return false;
            }

            T defaultObj = GetDefaultObjOfType<T>();

            switch (defaultObj)
            {
                #region Base Types
                //Misc
                case bool:
                case char:

                //Numbers
                case double:
                case float:
                case int:
                case long:
                case short:
                case uint:
                case ulong:
                case ushort:
                #endregion
                #region Derived Types
                //Vectors
                case Vector3:
                case Vector3Int:
                case Vector2:
                case Vector2Int:
                case Quaternion:
                    return true;
                #endregion
                default:
                    throw new NotImplementedException($"The size of {typeof(T)} cannot be found yet");
            }
        }

        public static void PutToBuffer(byte[] toBuffer, byte[] fromBuffer, ref int offset)
		{
            for (int i = 0; offset < fromBuffer.Length; offset++, i++)
			{
                toBuffer[offset] = fromBuffer[i];
			}
		}

        private static T GetDefaultObjOfType<T>()
        {
            if (typeof(T).IsValueType || typeof(T) == typeof(string))
            {
                return default(T);
            }
            else
            {
                return (T)Activator.CreateInstance(typeof(T));
            }
        }

        #region Debug
        public static string GetBytesAsString(IEnumerable<byte> bytesEnum)
        {
            string returnValue = string.Empty;
            byte[] bytes = bytesEnum.ToArray();
            int byteCount = bytes.Count();
            for (int i = 0; i < byteCount; i++)
            {
                returnValue += $"{bytes[i]}" + (i < byteCount - 1 ? "|" : "");
            }
            return returnValue;
        }

        public static byte[] GetStringAsBytes(string stringifiedBytes)
        {
            string[] splittedString = stringifiedBytes.Split('|');

            byte[] bytes = new byte[splittedString.Length];
            for ( int i = 0; i < splittedString.Length; i++)
            {
                bytes[i] = byte.Parse(splittedString[i]);
            }

            return bytes;
        }
        #endregion
    }
}

File: OnionMan_Unreal/Source/OnionMan_Unreal/Private/EncodingUtility.cpp
#include "EncodingUtility.h"
#include "Runtime/Core/Public/Containers/Array.h"
#include "Runtime/Core/Public/Math/Vector.h"
#include "Runtime/Core/Public/Math/Quat.h"
#include "Runtime/Core/Public/Math/IntVector.h"

namespace OnionMan::Network 
{
#pragma region PutEncodedValueInBuffer
    /*
    template<typename T>
    void EncodingUtility::PutEncodedValueInBuffer(T value, TArray<uint8>& buffer, int& offset)
    {
    }

    template<> 
    static void EncodingUtility::PutEncodedValueInBuffer<FString>(FString value, TArray<uint8>&buffer, int& offset)
    {
        PutToBuffer(buffer, (const uint8*)TCHAR_TO_UTF8(*value), offset, value.Len());
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<bool>(bool value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(bool));
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<char>(char value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(char));
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<double>(double value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(double));
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<float>(float value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(float));
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<int>(int value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(int));
    }

    template<> 
    static void EncodingUtility::PutEncodedValueInBuffer<long>(long value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(long));
    }

    template<> 
    static void EncodingUtility::PutEncodedValueInBuffer<short>(short value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(short));
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<uint32>(uint32 value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(uint32));
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<uint64>(uint64 value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(uint64));
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<uint16>(uint16 value, TArray<uint8>& buffer, int& offset)
    {
        PutToBuffer(buffer, (uint8*)&value, offset, sizeof(uint16));
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<FVector3f>(FVector3f value, TArray<uint8>& buffer, int& offset)
    {
        PutEncodedValueInBuffer(value.X, buffer, offset);
        PutEncodedValueInBuffer(value.Y, buffer, offset);
        PutEncodedValueInBuffer(value.Z, buffer, offset);
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<FIntVector>(FIntVector value, TArray<uint8>& buffer, int& offset)
    {
        PutEncodedValueInBuffer(value.X, buffer, offset);
        PutEncodedValueInBuffer(value.Y, buffer, offset);
        PutEncodedValueInBuffer(value.Z, buffer, offset);
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<FVector2f>(FVector2f value, TArray<uint8>& buffer, int& offset)
    {
        PutEncodedValueInBuffer(value.X, buffer, offset);
        PutEncodedValueInBuffer(value.Y, buffer, offset);
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<FIntVector2>(FIntVector2 value, TArray<uint8>& buffer, int& offset)
    {
        PutEncodedValueInBuffer(value.X, buffer, offset);
        PutEncodedValueInBuffer(value.Y, buffer, offset);
    }

    template<>
    static void EncodingUtility::PutEncodedValueInBuffer<FQuat4f>(FQuat4f value, TArray<uint8>& buffer, int& offset)
    {
        PutEncodedValueInBuffer(value.X, buffer, offset);
        PutEncodedValueInBuffer(value.Y, buffer, offset);
        PutEncodedValueInBuffer(value.Z, buffer, offset);
        PutEncodedValueInBuffer(value.W, buffer, offset);
    }
    */
#pragma endregion

#pragma region Encode
    /*
    template <typename T> 
    static void EncodingUtility::Encode(T value, TArray<uint8>& encodedValue)
    {
        throw "The type " << T << "cannot be encoded yet";
    }

    template<>
    static void EncodingUtility::Encode<FString>(FString value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (const uint8*)TCHAR_TO_UTF8(*value), value.Len() };
    }

    template<> 
    static void EncodingUtility::Encode<bool>(bool value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(bool) };
    }

    template<> 
    static void EncodingUtility::Encode<char>(char value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(char) };
    }

    template<> 
    static void EncodingUtility::Encode<double>(double value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(double) };
    }

    template<> 
    static void EncodingUtility::Encode<float>(float value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(float) };
    }

    template<> 
    static void EncodingUtility::Encode<int>(int value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(int) };
    }

    template<>
    static void EncodingUtility::Encode<long>(long value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(long) };
    }

    template<>
    static void EncodingUtility::Encode<short>(short value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(short) };
    }

    template<>
    static void EncodingUtility::Encode<uint32>(uint32 value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(uint32) };
    }

    template<>
    static void EncodingUtility::Encode<uint64>(uint64 value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(uint64) };
    }

    template<>
    static void EncodingUtility::Encode<uint16>(uint16 value, TArray<uint8>& encodedValue)
    {
        encodedValue = TArray<uint8>{ (uint8*)&value, sizeof(uint16) };
    }

    template<>
    static void EncodingUtility::Encode<FVector3f>(FVector3f value, TArray<uint8>& encodedValue)
    {
        encodedValue.Empty();
        encodedValue.Reserve(3 * sizeof(float));

        TArray<uint8> encodedElement{};

        Encode(value.X, encodedElement);
        encodedValue.Append(encodedElement);

        Encode(value.Y, encodedElement);
        encodedValue.Append(encodedElement);

        Encode(value.Z, encodedElement);
        encodedValue.Append(encodedElement);
    }

    template<>
    static void EncodingUtility::Encode<FIntVector>(FIntVector value, TArray<uint8>& encodedValue)
    {
        encodedValue.Empty();
        encodedValue.Reserve(3 * sizeof(int));

        TArray<uint8> encodedElement{};

        Encode(value.X, encodedElement);
        encodedValue.Append(encodedElement);

        Encode(value.Y, encodedElement);
        encodedValue.Append(encodedElement);

        Encode(value.Z, encodedElement);
        encodedValue.Append(encodedElement);
    }

    template<>
    static void EncodingUtility::Encode<FVector2f>(FVector2f value, TArray<uint8>& encodedValue)
    {
        encodedValue.Empty();
        encodedValue.Reserve(2 * sizeof(float));

        TArray<uint8> encodedElement{};

        Encode(value.X, encodedElement);
        encodedValue.Append(encodedElement);

        Encode(value.Y, encodedElement);
        encodedValue.Append(encodedElement);
    }

    template<>
    static void EncodingUtility::Encode<FIntVector2>(FIntVector2 value, TArray<uint8>& encodedValue)
    {
        encodedValue.Empty();
        encodedValue.Reserve(2 * sizeof(int));

        TArray<uint8> encodedElement{};

        Encode(value.X, encodedElement);
        encodedValue.Append(encodedElement);

        Encode(value.Y, encodedElement);
        encodedValue.Append(encodedElement);
    }

    template<>
    static void EncodingUtility::Encode<FQuat4f>(FQuat4f value, TArray<uint8>& encodedValue)
    {
        encodedValue.Empty();
        encodedValue.Reserve(4 * sizeof(float));

        TArray<uint8> encodedElement{};

        Encode(value.X, encodedElement);
        encodedValue.Append(encodedElement);

        Encode(value.Y, encodedElement);
        encodedValue.Append(encodedElement);

        Encode(value.Z, encodedElement);
        encodedValue.Append(encodedElement);

        Encode(value.W, encodedElement);
        encodedValue.Append(encodedElement);
    }
    */
#pragma endregion

#pragma region Decode
/*
    template <typename T>
    static T EncodingUtility::Decode(TArray<uint8> bytes, int& offset, int size)
    {
        throw "The type " << T << "cannot be decoded yet";
    }

    template<>
    static FString EncodingUtility::Decode<FString>(TArray<uint8> bytes, int& offset, int size)
    {
        if (size == -1)
        {
            throw "In order to decode a string, the argument 'size' must be set";
        }
        FString decodedString = FString(UTF8_TO_TCHAR(bytes.GetData() + offset));
        offset += size;
        return decodedString;
    }

    template<>
    static bool EncodingUtility::Decode<bool>(TArray<uint8> bytes, int& offset, int size)
    {
        bool decodedValue = *((bool*)(bytes.GetData() + offset));
        offset += sizeof(bool);
        return decodedValue;
    }

    template<>
    static char EncodingUtility::Decode<char>(TArray<uint8> bytes, int& offset, int size)
    {
        char decodedValue = *((char*)(bytes.GetData() + offset));
        offset += sizeof(char);
        return decodedValue;
    }

    template<>
    static double EncodingUtility::Decode<double>(TArray<uint8> bytes, int& offset, int size)
    {
        double decodedValue = *((double*)(bytes.GetData() + offset));
        offset += sizeof(double);
        return decodedValue;
    }

    template<>
    static float EncodingUtility::Decode<float>(TArray<uint8> bytes, int& offset, int size)
    {
        float decodedValue = *((float*)(bytes.GetData() + offset));
        offset += sizeof(float);
        return decodedValue;
    }

    template<>
    static int EncodingUtility::Decode<int>(TArray<uint8> bytes, int& offset, int size)
    {
        int decodedValue = *((int*)(bytes.GetData() + offset));
        offset += sizeof(int);
        return decodedValue;
    }

    template<>
    static long EncodingUtility::Decode<long>(TArray<uint8> bytes, int& offset, int size)
    {
        long decodedValue = *((long*)(bytes.GetData() + offset));
        offset += sizeof(long);
        return decodedValue;
    }

    template<>
    static short EncodingUtility::Decode<short>(TArray<uint8> bytes, int& offset, int size)
    {
        short decodedValue = *((short*)(bytes.GetData() + offset));
        offset += sizeof(short);
        return decodedValue;
    }

    template<>
    static uint32 EncodingUtility::Decode<uint32>(TArray<uint8> bytes, int& offset, int size)
    {
        uint32 decodedValue = *((uint32*)(bytes.GetData() + offset));
        offset += sizeof(uint32);
        return decodedValue;
    }

    template<>
    static uint64 EncodingUtility::Decode<uint64>(TArray<uint8> bytes, int& offset, int size)
    {
        uint64 decodedValue = *((uint64*)(bytes.GetData() + offset));
        offset += sizeof(uint64);
        return decodedValue;
    }

    template<>
    static uint16 EncodingUtility::Decode<uint16>(TArray<uint8> bytes, int& offset, int size)
    {
        uint16 decodedValue = *((uint16*)(bytes.GetData() + offset));
        offset += sizeof(uint16);
        return decodedValue;
    }

    template<>
    static FVector3f EncodingUtility::Decode<FVector3f>(TArray<uint8> bytes, int& offset, int size)
    {
        return FVector3f(
            Decode<float>(bytes, offset),
            Decode<float>(bytes, offset),
            Decode<float>(bytes, offset));
    }

    template<>
    static FIntVector EncodingUtility::Decode<FIntVector>(TArray<uint8> bytes, int& offset, int size)
    {
        return FIntVector(
            Decode<int>(bytes, offset),
            Decode<int>(bytes, offset),
            Decode<int>(bytes, offset));
    }

    template<>
    static FVector2f EncodingUtility::Decode<FVector2f>(TArray<uint8> bytes, int& offset, int size)
    {
        return FVector2f(
            Decode<float>(bytes, offset),
            Decode<float>(bytes, offset));
    }

    template<>
    static FIntVector2 EncodingUtility::Decode<FIntVector2>(TArray<uint8> bytes, int& offset, int size)
    {
        return FIntVector2(
            Decode<int>(bytes, offset),
            Decode<int>(bytes, offset));
    }

    template<>
    static FQuat4f EncodingUtility::Decode<FQuat4f>(TArray<uint8> bytes, int& offset, int size)
    {
        return FQuat4f(
            Decode<float>(bytes, offset),
            Decode<float>(bytes, offset),
            Decode<float>(bytes, offset),
            Decode<float>(bytes, offset));
    }
    */
#pragma endregion

#pragma region PutToBuffer
    void EncodingUtility::PutToBuffer(TArray<uint8>& toBuffer, TArray<uint8>& fromBuffer, int& offset)
    {
        for (int i{ 0 }; i < fromBuffer.Num(); i++, offset++)
        {
            toBuffer[offset] = fromBuffer[i];
        }
    }

    void EncodingUtility::PutToBuffer(TArray<uint8>& toBuffer, const uint8* fromBuffer, int& offset, int fromBufferSize)
    {
        for (int i{ 0 }; i < fromBufferSize; i++, offset++)
        {
            toBuffer[offset] = fromBuffer[i];
        }
    }
#pragma endregion
/*
#pragma region GetSizeOf

    template<typename T>
    static int EncodingUtility::GetSizeOf(T* value)
    {
        return sizeof(T);
    }

    template<>
    static int EncodingUtility::GetSizeOf<FString>(FString* value)
    {
        TArray<uint8> encodedString;
        int offset = 0;
        PutEncodedValueInBuffer(*value, encodedString, offset);
        return encodedString.Num();
    }

    template<>
    static int EncodingUtility::GetSizeOf<FVector3f>(FVector3f* value)
    {
        return 3 * sizeof(float);
    }

    template<>
    static int EncodingUtility::GetSizeOf<FIntVector>(FIntVector* value)
    {
        return 3 * sizeof(int);
    }

    template<>
    static int EncodingUtility::GetSizeOf<FVector2f>(FVector2f* value)
    {
        return 2 * sizeof(float);
    }

    template<>
    static int EncodingUtility::GetSizeOf<FIntVector2>(FIntVector2* value)
    {
        return 2 * sizeof(int);
    }

    template<>
    static int EncodingUtility::GetSizeOf<FQuat4f>(FQuat4f* value)
    {
        return 4 * sizeof(float);
    }

#pragma endregion
*/
/*
#pragma region HasFixedEncodedSize
    template<typename T>
    static bool EncodingUtility::HasFixedEncodedSize()
    {
        return true;
    }

    template<>
    static bool EncodingUtility::HasFixedEncodedSize<FString>()
    {
        return false;
    }
    */
#pragma endregion

#pragma region Debug
    FString EncodingUtility::GetBytesAsString(const TArray<uint8>& bytes)
    {
        FString returnValue;
        int byteCount = bytes.Num();
        for (int i = 0; i < byteCount; i++)
        {
            returnValue += FString::FromInt(bytes[i]) + (i < byteCount - 1 ? "|" : "");
        }
        return returnValue;
    }

    void EncodingUtility::GetStringAsBytes(FString stringifiedBytes, TArray<uint8>& bytes)
    {
        TArray<FString> splittedString;
        stringifiedBytes.ParseIntoArray(splittedString, TEXT("|"), true);

        bytes.Empty();
        bytes.SetNumUninitialized(splittedString.Num());
        for (int i = 0; i < splittedString.Num(); i++)
        {
            bytes[i] = FCString::Atoi(*splittedString[i]);
        }
    }
#pragma endregion
}